### DISCUSSION TITLE
Welcome to SmartSpin2k Discussions!

### DISCUSSION THREAD

**Discussion by doudar** [doudar/SmartSpin2k#37] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by fireyeti** [comment]
Hi All-

I'm excited to be part of the group. I'm Nick from the Detroit area. I spend most of my riding time on my mountain bike on a local trail. My road bike is approaching historical status, but still carries me around the nice local gravel roads. I have a RevMaster spin bike on which I plan to install the SmartSpin2k during the cold weather and for travel purposes.

I hope to develop the community with install, doc, and testing help. Still interested in learning code, so following along the development of the software!

-fireyeti

**Reply by doudar** [reply]
Hi Nick, 

Thanks for contributing the photos of the RevMaster spin so we can create the appropriate adapter for it!

**Reply by doudar** [reply]
> Hi All-
> 
> I'm excited to be part of the group. I'm Nick from the Detroit area. I spend most of my riding time on my mountain bike on a local trail. My road bike is approaching historical status, but still carries me around the nice local gravel roads. I have a RevMaster spin bike on which I plan to install the SmartSpin2k during the cold weather and for travel purposes.
> 
> I hope to develop the community with install, doc, and testing help. Still interested in learning code, so following along the development of the software!
> 
> -fireyeti

Done testing Nick! Ready to ship it out to you tomorrow!
![B6F96CD0-A023-4FB7-9A16-7C26979B46E0](https://user-images.githubusercontent.com/17362216/102730457-74934d80-42fa-11eb-8577-ed78aa15bf7f.jpeg)
![C0D45591-9638-4DA2-81C2-330F1A01237E](https://user-images.githubusercontent.com/17362216/102730460-778e3e00-42fa-11eb-8957-94262f6ba2c8.jpeg)

**Reply by fireyeti** [reply]
Can't wait to try it out! I have a short term power solution in HR2VP which seems to be fairly accurate, just slow response to effort so not ideal for racing. Pedals are on my list!

**Reply by doudar** [reply]
I was actually thinking that functionality is something that could be added to this in the future since it has the ability to be paired to a HRM in addition to a power meter. Laggy, but yes, it would be better than nothing. 

Take lots of pictures and screenshots when you get it....I can't wait to see that documentation. You have no idea what you may have just gotten yourself into...lol

**Reply by fireyeti** [reply]
Definitely could be added. I don't know how tricky it would be to code that algorithm.

Don't worry when I submit my horrible pictures you'll wish I never found your project!

![0B1BED6B-F87B-458B-A28D-954272D9A0E8](https://user-images.githubusercontent.com/76080977/102733446-58e47300-430b-11eb-9a23-4b4ffba78128.jpeg)

**Reply by doudar** [reply]
Your (completely assembled) kit was shipped. BTW, 

It's configured to grab any unconnected power meter or heart monitor on powerup, so if you'd like to pair those to something else, do so before you turn on the smart spin. Another little tidbit of information is the blue light provides some limited indications. A slow steady blink means it's waiting for a BLE client connection, steady on means it's connected, very fast flashing means it's updating (from github).

**Comment by sspieser** [comment]
Hello, I'm from France, did a lot of trail running but my knees are now out of order... so switching to cycling ! I have a BH SB3 spinbike + PowerTap P1 (also have Assioma DUA if required...), and I would enjoy building that kind of remote! The tricky part is for me the hardware (3D printing...), but I hope to find the time to join and will contribute if I can help... As of today, I have an ESP32 I have never used :-)

**Reply by doudar** [reply]
Hello @sspieser 

If you take a look in the Hardware/Inserts/Design_Reference_Photos folder, you’ll see a couple images that @fireyeti supplied. If you could post the same measurements (photos like he did are best), I’ll add them to the archive and design the appropriate knob adapter for you to print. 

As far as printing goes, maybe we can get some of the STL’s uploaded to one of the online print manufacturers so that people could easily grab/order parts from there. I’ve never used one of the online printing services but there are a lot of options nowadays.

**Reply by doudar** [reply]
Also, any BLE power meter should work. It’s also only required to be paired with the SmartSpin2k for ERG mode. So if there is a power meter that is incompatible for any reason, SIm mode will still work (grade changes from the app will work).

**Reply by fireyeti** [reply]
@sspieser https://github.com/doudar/SmartSpin2k/tree/master/Hardware/Inserts/Design_Reference_Photos

**Reply by sspieser** [reply]
I'll take pictures during the vacations 👍 ... also will try to order the electronic parts to see the stepper running ;-)

**Reply by sspieser** [reply]
> If you could post

Hi, how can I upload the pictures? Seems I don't have write acces to add a file... thx

**Reply by doudar** [reply]
You can actually just drag and drop them here in the discussion window and I'll add them. 

Thanks!

**Reply by sspieser** [reply]
![BH-SB3Magnetic_TopView](https://user-images.githubusercontent.com/14963451/102824085-703f6100-43dc-11eb-9989-56aec2c0ed7d.jpg)
![BH-SB3Magnetic_SideView](https://user-images.githubusercontent.com/14963451/102824103-76354200-43dc-11eb-9b67-1abf2a393f28.jpg)
![BH-SB3Magnetic_Sidedraw_cotation](https://user-images.githubusercontent.com/14963451/102824132-864d2180-43dc-11eb-97ff-5bbdb165a594.jpg)

**Reply by sspieser** [reply]
Hope this is clear enough ?

**Reply by doudar** [reply]
Looks great! 

I'll overlay your drawing quick and then upload it. 

Thank you!

**Reply by fireyeti** [reply]
Hi @sspieser , the knob is not an even sided object, and I'm not certain where your exact measurements overlay the photograph. Can you confirm the measurements across the distances I marked in this photo? The schematic is excellent otherwise!
![Top View](https://user-images.githubusercontent.com/76080977/102900829-d955c180-443a-11eb-92d5-ecb79269d8d4.jpg)

**Comment by sspieser** [comment]
Hello!
The measurement as your draw is 55mm !

Today I received the LM2596, things are going concrete...
Thanks a lot, Seb

**Reply by fireyeti** [reply]
Got it. Thanks! Files will be uploaded for @doudar shortly.

**Comment by drz01** [comment]
Charles from Melbourne here. Got a spin bike after discovering HR2VP, and came to this project via your comments after Ray's article. I'm most excited by the concept, but I believe SS2k should similarly offer estimated power from cadence, which would raise its popularity, especially since progress in pedal meters has been glacial.

Further hardware simplication will be needed for a broad appeal. The PCB kit idea does sound great. But until there is a full database of bike models with associated "print to order" inserts, I wonder if silicone putty mods (e.g. Sugru) will be easier, as the standard kit can include a preprinted basic insert.

I aim to contribute where I can in the coming weeks and months. Thanks all!

**Comment by doudar** [comment]
Hi Charles, glad to have you! 

I'm phasing in a new larger knob cup (V2) which fits over the top of all knobs we've encountered so far. With that version, using Surgru would work very well. It's a great idea! 

Check out [Knob_Cup_Untested_Large.STL](https://github.com/doudar/SmartSpin2k/blob/master/Hardware/Knob_Cup_Untested_Large.STL), into which the Sugru could possibly be directly molded. If the Sugru is permanent (I've never worked with it), or you need more texture, it would be easy to use an insert like [Untested_Revmaster_Spin_For_Large_Cup.STL](https://github.com/doudar/SmartSpin2k/blob/master/Hardware/Inserts/Untested_Revmaster_Spin_For_Large_Cup.STL) to mold the Sugru into. If you need any other insert designs (more facets for texture?), let me know. I'm getting pretty good at whipping them up quickly. 

The power from cadence is doable, but it's a pretty big multi-step project. We would need a few power readings (or estimates) across the resistance range at a few different cadences and then interpolate between them.  Possibly this could be automatically trained with something like HR2VP, or a borrowed power source and then stored in a lookup table in spiffs. We'd also have to implement homing at each powerup (not a big issue, the stepper drivers already support it). I'm definitely open to it

In the mean time, @fireyeti has reported that the SS2K actually works pretty well with HR2VP, even completing ERG workouts with it. I've been toying with the idea of adding in our own sort of HR2VP functionality. All of the basic elements are already in place, we just have to figure out the inputs needed and the best way to calculate it. 

Let me know if you need any help or would like me to put a kit together for you. Australia is a ways away so shipping could be expensive & long but I've got plenty of parts here and I'm sure we could work something out.

**Reply by drz01** [reply]
Thank you for the very encouraging feedback. I'm thinking of starting with a POC comparison of a custom print and Sugru mod.

One issue I've already seen is in ordering components; local availability or shortages will lead to alternate suppliers; the variants might then not fit with the printed parts?? I'll be most glad to reimburse you for your help; though we should definitely set up a more sustainable solution for everyone.

I think even integrating HR2VP-like functionality will be a major improvement over using the app; in iOS it has to stay in the foreground and therefore will lock out the device for the entire workout! This already sounds like a side project to which I can contribute.

Back to work in not many hours! More on all of the above later this week.

**Reply by doudar** [reply]
The only part I’ve seen a lot of variance on is the stepper body length, but as long as it’s nema 17 and less than 38mm (I think that’s the number), it fits under the cover. The ESP32 devkit is a standard, so are the stepper drivers.

What might help with the international crowd (let me know, I’m just talking out loud) is if I put links to the stuff on Aliexpress. I can put that on my short term to-do list. 

If you’re willing & able to start working on a heart rate to power function, I’d love it!!! Feel free to fork the project and mess around with it. All you need to start working on it is just the esp32 devkit. 

As a short head start, you can get the current heart rate from userConfig.getSimulatedHeartRate() and set the power with userConfig.setSimulatedHeartRate(powerInWatts)

**Reply by doudar** [reply]
Also, thinking about it, it would probably be worthwhile to explore GitHub for a HR2VP like app. I be someone has already figured out the bones of that problem.

**Reply by doudar** [reply]
Here’s a great resource:
https://www.braydenwm.com/calburn.htm#:~:text=You%20can%20increase%20the%20accuracy,%2Fhour)*1000%2F3600.

**Comment by lenzirodrigo** [comment]
Hello Anthony, 
I found this initiative mentioned on the [r/SchwinnIC4_BowflexC6](https://www.reddit.com/r/SchwinnIC4_BowflexC6/) reddit sub.
Looks awesome and I would like to build it for my SchwinnIC8 bike, I'm also willing to support the development, unfortunately Im not versed with C++, nor arduino, willing to learn though.
I will be shopping for the parts in the following weeks and set it up to start testing it. 

The [initial set up wiki](https://github.com/doudar/SmartSpin2k/wiki/Initial-Set-Up) mentions a build guide that I was not able to find, would you mind sharing it to me.

Thanks!
Rodrigo

**Reply by doudar** [reply]
Unfortunately, most of what we have for build documentation is currently on our main readme.md and helped a bit with the exploded cad video. It's the area of the project that currently needs the most help. 

Every time I personally start on documentation, other areas of the project demand my attention.  

This week I'm sending some parts to @fireyeti so he can document building the non-PCB version. That should help a lot, and he's much better with the documentation aspect than I am. 

Feel free to post as much as needed here in the discussion and we will all try to help you out as best as possible. The discussion should help others with their builds as well. Once you start work on it, it would be very cool if you started a discussion topic specifically for your build with progress pictures and such. 

Good Luck! I can't wait to see more!

**Reply by lenzirodrigo** [reply]
Hey Anthony,

I was looking for the components and Amazon.es for electronics lacks of items or prices are super high, so I end up checking on aliexpress. 
Would you mind double checking the list before I order?

| qty |           Part          | Link                    | aliexpress                                                                                   |   |
|-----|:-----------------------:|-------------------------|----------------------------------------------------------------------------------------------|---|
| 1   | ESP32 Dev Board         | https://amzn.to/2ZNyjQX | https://www.aliexpress.com/item/1005001267643044.html |   |
| 1   | NEMA 17 Pancake Stepper | https://amzn.to/37mKKHh | https://www.aliexpress.com/item/4000120288783.html    | I did not find 34mm, its 38mm will it fit?  |
| 1   | TMC2225                 | https://amzn.to/3kctdEQ | https://www.aliexpress.com/item/4001149124672.html    |   |
| 1   | LM2596 Buck Converter   | https://amzn.to/33ofggY | https://www.aliexpress.com/item/33014890815.html      |   |
| 1   | 24V Power Supply        | https://amzn.to/3r4e1i0 | https://www.aliexpress.com/item/32975192317.html      |   |
| 2   | Tactile Switches        | https://amzn.to/33ezmKx | https://www.aliexpress.com/item/32958087576.html      |   |
| 2   | 608 Skate Bearings      | https://amzn.to/3isBzrW | https://www.aliexpress.com/item/32700232097.html      |   |

What is the name of the socket to connect the power supply? I'm refering to the one on the first picture ([IMG_7236.jpg](https://github.com/doudar/SmartSpin2k/blob/master/Pictures/Assembly/IMG_7236.jpg))I could not find any online as I do not know the name :) 
Also do you see anything that I might miss? I have iron solder as I built a mechanical keyboard, but if you have suggestions for the tools please advise.

Regarding the performance of SmartSpin2k, what is the latency or time to react and adjust the resistance on Zwift? I understand that code can be improved,(and hope to contribute) but do you face limitations on the electronic components? I'm thinking on the stepping motor.

Thanks!

**Reply by doudar** [reply]
`Regarding the performance of SmartSpin2k, what is the latency or time to react and adjust the resistance on Zwift? I understand that code can be improved,(and hope to contribute) but do you face limitations on the electronic components? I'm thinking on the stepping motor.`

Response time is great. We only delay 700ms (worst case) from the time a BLE signal is received via a sensor until it is processed and forwarded to the app. Most power meters take about that much time to analyze so it seems negligible to me. The resistance changes received from Zwift are received and processed very quickly. 

The SmartSpin2k isn't a "torque monster". It provides a reasonable amount of torque and should move your knob from stop to stop on a magnetic bike and at least most of the way on friction based resistance bikes. You might have to clean and lubricate your bikes mechanism or loosen a tension adjustment to get optimum travel out of it. We did find one bike recently where the performance was unsatisfactory because the knob only required 3/4 of a turn to go from minimum resistance to maximum resistance and that just isn't enough mechanical advantage for this unit.

**Reply by fireyeti** [reply]
Hello Rodrigo! 

Thank you again for the AliExpress list, it will make things simpler for our international users!

I have finshed the Instructable which @doudar set up and it's published [here](https://www.instructables.com/Spin-Bike-Controller/). Please let us know if you have any comments our would like additional clarification. As @doudar mentioned I'm planning to document a build of a non PCB board so hopefully there will be even more helpful information avaiable soon.

Welcome and we look forward to your contributions!

Nick

**Reply by lenzirodrigo** [reply]
> Response time is great. We only delay 700ms (worst case) from the time a BLE signal is received via a sensor until it is processed and forwarded to the app. Most power meters take about that much time to analyze so it seems negligible to me. The resistance changes received from Zwift are received and processed very quickly.
> 
> The SmartSpin2k isn't a "torque monster". It provides a reasonable amount of torque and should move your knob from stop to stop on a magnetic bike and at least most of the way on friction based resistance bikes. You might have to clean and lubricate your bikes mechanism or loosen a tension adjustment to get optimum travel out of it. We did find one bike recently where the performance was unsatisfactory because the knob only required 3/4 of a turn to go from minimum resistance to maximum resistance and that just isn't enough mechanical advantage for this unit.

Good to know this, I'm more confident to speed up the build. My spin bike (Schiwinn IC8 - EU version of the IC4) has several turns to go from min to max. I'm not planning to keep the zwift subscription as I plan to ride outside with the good weather, so I need to hurry up :)

Thanks for the details!

**Comment by doudar** [comment]
Esp32 is good - don't get the "yellow" option. Get the one with the fewer pins. 

I'd recommend the BigTreeTech stepper driver boards. The other brands sometimes have their UART jumpers in different configurations and that can take additional troubleshooting time to sort out. https://www.aliexpress.com/item/4000296898203.html

The 38mm stepper motor is good. I think that's actually the size I'm currently using. I'll have to update the readme.md . The only issue you may run into is if it is too long, you might need to make or modify the top cover to be a bit deeper. 

In addition, the "optional" stuff you'll need is:

3.5mm stereo headphone connector (shifter connector) https://www.aliexpress.com/item/4000640677390.html

5.5X2.1mm Dc Power connector https://www.aliexpress.com/item/4000694128319.html

Stereo RCA-->3.5mm headphone "Y" Cable (Easier and cheaper than making my own "Y" Cables - cut off the RCA ends and attach the shifter buttons to them). https://www.aliexpress.com/item/4000204275028.html

3mm screws to attach the stepper motor to the crossbar and case. 

4 #6 X 3/4" (stainless is best) wood screws to attach the top cover. 

2 #4 X 1/2" flat head screws to attach the ESP32 . 

Hot Glue to attach the access plugs to the shifter straps and to fix the TMC2225 in place.

**Reply by marklogan17** [reply]
Quick question about screws to attach the buck converter.  I started by attempting to use the #4 X 1/2 and they were too big.  I improvised a little with some old filament and hot glue (kinda made a rivet out of it), but I figured I would ask what everyone else is using as I'm building a couple more of these for friends and would prefer to screw it in if possible.

**Reply by doudar** [reply]
I've used some #4 or #6 x 1/4 screws on that before. Honestly it's untracked materials out of my screw bins whenever I've built the non-pcb version. I've also cut screws to length in the past as the plastic is pretty forgiving to malformed screw threads as long as the diameter isn't too big. 

Let me know if you need any edits to those mounts as it's been a while since I've done any updates to them. The PCB version just deletes the mounts so I don't look at them much as I've been more focused on the PCB version. 

Sounds like fun! Post some pictures of the progress when you can!

**Reply by marklogan17** [reply]
I think the mounts themselves are fine.  The ESP one went in no problem.  The TMC2225 fits fine and the glue should hold it.  It was just the buck converter one was a little short for the 1/2" screw.  You couldn't make it longer as it would interfere with the ESP board pins too much if its further out.  As it is I had to slightly bend the gnd pin that the shifter is connected to to fit it in there.  I'll take some pictures after I get the TMC2225 glued in and I confirm I haven't actually broken anything while mounting in the box :)

**Reply by doudar** [reply]
> I think the mounts themselves are fine.  The ESP one went in no problem.  The TMC2225 fits fine and the glue should hold it.  It was just the buck converter one was a little short for the 1/2" screw.  You couldn't make it longer as it would interfere with the ESP board pins too much if its further out.  As it is I had to slightly bend the gnd pin that the shifter is connected to to fit it in there.  I'll take some pictures after I get the TMC2225 glued in and I confirm I haven't actually broken anything while mounting in the box :)

Quick note, I've recently noticed the the holes in the stepper crossbar are oversized  just enough that it can get installed on the stepper improperly and provide enough room for the gears to slip when under load. This is fixed upstream by a 31 tooth gear that I'll put into the repo soon. 

So, the proper way to install the crossbar onto the motor is to make sure it is as close to the stepper axle as possible before tightening the two m3x8 screws. 

Also, best way to install the small gear onto the stepper shaft is to heat the gear slightly with a heat gun and it will then slide right on and lock in place when it cools.

Pounding works too but probably hard on the motor bearings and occasionally a gear will split 😬.

**Reply by doudar** [reply]
Btw, here is the 31t gear incase you want it for your future builds. It's a little tighter meshing so you may need to adjust the crossbar opposite of the above procedure if you use it:

[Spur gear 1M 31T.zip](https://github.com/doudar/SmartSpin2k/files/6448336/Spur.gear.1M.31T.zip)

**Comment by urb4nc4rl** [comment]
I've just begun my SmartSpin2K journey!

I'm a software developer, both back-end and also mobile.  I'm getting more into indoor cycling as a result of the pandemic and trying out both 'Zwift' and also Peloton Digital; my main motivation for joining this project is that there is no easy way to calibrate the absolute resistance on my spin bike; I'd like a reproducible way of setting an absolute resistance rather than reducing this to guesswork.

I've already begun writing an iPhone companion app to aid in this quest, for displaying the resistance and also (hopefully) simulating 'Shift Up' / 'Shift Down' programatically over BLE.. ..have put some ideas into the discussion forum to see what we can make happen!

**Comment by chrisbmcarthur** [comment]
A quick thanks! I was using a Saris M2 trainer and swapping between 2 bikes for my spouse and myself. My wife wanted a spin bike for the feel/resistance it provides. I found your project, invested in a Schwinn IC4 and am now a SS2K v2 user. Thanks to eMadMan for providing the raw parts and guidance as required!
Your design has really matured and you have worked simplicity into the design. The SW is simple to get operational...really nicely done.
Once I have a thousand kms under my belt on this system I will see if there is any feedback/improvements I can offer. All the best. CMc

PS - 2 coworkers are also running your SW as a BT hub when dealing with Apple TV. There is an "Idea" created requesting that a discrete Cadence sensor configurable and combined into the stream to Zwift. We are working on modifications to the code ourselves to do such...a bit more work to go. I think that there is a complete parallel market for a flexible BT/BLE multiplexer....which is part of what you have created.

**Reply by doudar** [reply]
Awesome! That's great to hear!

Btw, we've just merged a major commit into develop if you want to check it out. I'd love it if you tested the binary on your ic4 before I push the OTA update. 

I'm always looking for more contributors so if you're team has anything to push back into this repo, I'm open to almost anything.

Have a great day!

**Reply by chrisbmcarthur** [reply]
Loaded the build from today - the Debug page is not showing the same info
![image](https://user-images.githubusercontent.com/80929246/213937222-d9957b80-e952-45fb-b42d-c4670dce1bb8.png)
In Zwift the HR/Cadence/Power is shown.
this is the 23.1.22 FW. I was previously on 23.1.11 and details were always present in these fields

**Reply by doudar** [reply]
Sorry, I forgot to mention some the html files were updated as well. You could upload those manually (from the data folder) or reset to defaults and it will update them from the develop branch.


Otherwise is it working correctly?

**Reply by doudar** [reply]
You can use the firmware update link on the webpage. Username and pass are both admin. Then individually upload each .html file in this folder. There's no need to reload the page between sending these files either:

https://github.com/doudar/SmartSpin2k/tree/develop/data

**Reply by doudar** [reply]
Just be careful not to upload anything not in the data folder as it will rapidly fill up the file system on the esp32 and you will have to reset to defaults to wipe it clean again.

**Comment by chrisbmcarthur** [comment]
Can you provide more detail on uploading manually please?

**Comment by chrisbmcarthur** [comment]
Not having fun ;-> Uploaded the 13 files in the github directory provided.
REbooted after. White screens for all html pages. OTA page works - that is
the only one I can manually access.
I need to revert ASAP for an upcoming ride....

**Reply by chrisbmcarthur** [reply]
Root cause - i attempted to download each file from GitHib via Save as.....and I unfortunately get the complete Github page. 
I will have to factory reset the ESP to clear all the files and upload...as the OTA page is now failing too.

**Reply by doudar** [reply]
Yah, if you upload anything with a different name than what's already on the file system, there's no way to remove it and if the file system is full, we can't write to it.

The good news is the reset is pretty painless. Either do it from the settings page or hold both shifters during boot.

**Comment by chrisbmcarthur** [comment]
I cannot get a settings page - the html file is invalid.
I tried the 2 button hold down - now all settings are gone and even when on
th eSmartSpin2k SSID all the html files are invalid.
I must do a full factory reset of the device - clear all storage - and
build up again.
Note - the Troubleshooting/Factory Reset page on teh wiki - the image
showing the reset pins is missing -
https://github.com/doudar/SmartSpin2k/wiki/Loading-Software
Could you update the pic some time for clarity? Thanks!

CMc

**Reply by eMadman** [reply]
While the device is powered off, press and hold both shifters.  Plug in the power cable.  Wait 2-3 seconds and release the shifter buttons.  This should force a reset of the firmware.  Try that before going down the path of flashing over USB.

If you need to go by USB, it's the Boot pin on the esp32 that needs to be pressed.  Look at the front of the case, to the left of the bike mount - you'll see two holes.  Lower one is the boot pin, upper one is EN.

**Comment by chrisbmcarthur** [comment]
Will the process you describe restore the html files from a known archive?
If not...it will not work because I pooched the html files.

**Reply by doudar** [reply]
> Will the process you describe restore the html files from a known archive? If not...it will not work because I pooched the html files.
> […](#)
> On Sun., Jan. 22, 2023, 6:02 p.m. eMadman, ***@***.***> wrote: While the device is powered off, press and hold both shifters. Plug in the power cable. Wait 2-3 seconds and release the shifter buttons. This should force a reset of the firmware. Try that before going down the path of flashing over USB. If you need to go by USB, it's the Boot pin on the esp32 that needs to be pressed. Look at the front of the case, to the left of the bike mount - you'll see two holes. Lower one is the boot pin, upper one is EN. — Reply to this email directly, view it on GitHub <[#37 (reply in thread)](https://github.com/doudar/SmartSpin2k/discussions/37#discussioncomment-4751793)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ATJODXU6F3KCHPUKVUN577DWTW4AFANCNFSM4VABFA6A> . You are receiving this because you commented.Message ID: ***@***.***>

Any time you reset to defaults, the partition that holds the html files is erased. On the next boot, you will have to connect to the SmartSpin2k SSID and then on the webpage that shows up, enter the network credentials for your 2.4ghz network. Then hit save and then reboot. On the next boot, the SS2K will connect to your home network and automatically download the html files from github.

**Reply by eMadman** [reply]
It should bring you back to a first boot state and ask you to connect to your home wifi network.  After that, it will re-download the latest html files.

Flashing by USB will bring it back to the same state

Load the firmware.bin into smartspin2k flasher and select your esp32
![image](https://user-images.githubusercontent.com/24726844/213946324-9128568a-58b3-4897-b0ad-8b2f387f9a56.png)

Click Flash ESP
Press and hold boot pin.  Release after you see flashing start (approx 3-5s)
If that doesn't work, restart the flasher utility and repeat - this time, tap the EN button while holding boot pin

**Comment by chrisbmcarthur** [comment]
I follow the procedure, holding the shifter buttons for 3 seconds then
release. I know the buttons work because after the full boot, the motor
responds to shifts.
I goto the smartspin2k SSID. I receive a blank page. Here is the html
debug. Same situation if i modify the URL to settings.html

I attempted the boot button on the ESP32 - still comes up with SS2K FW and
still no html content.

[image: image.png]

**Reply by doudar** [reply]
After reviewing code - it looks like holding the shifters doesn't erase the partition, it only resets the settings - I'll add an issue for that. 

So to reset your partitions, you could either load via USB like @eMadman mentioned above, or

Connect to the SS2K on the SmartSpin2k SSID, open a web browser and go to http://192.168.4.1/load_defaults.html

Just loading that page will reset it to defaults and erase the partition. Remember to close that tab then otherwise your browser might try to intermittently reset your SS2K :)

**Comment by chrisbmcarthur** [comment]
I am able to flash the FW.....but the html pages are still blank.
 (as stated before, I placed the wrong html files on teh device)
I would like more information on the EN button, the Boot button and what
each does....searching....

**Reply by chrisbmcarthur** [reply]
Well....troubleshooting is over for a while...the device fell over while the USB was plugged in and ripped the USB jack off the PCB. Sigh.....all from attempting to load a fresh build and provide feedback, :->
Thanks for the attempts to assist.

**Reply by doudar** [reply]
Yikes! Well at least you did find a bug on the shifter reset....

The good news is that if you have a working firmware onboard, you should be able to reset it using this link http://192.168.4.1/load_defaults.html like above ^^
Obviously you'll want to make sure no other traces are damaged on the devkit before you power it up. 

You'll rarely use the USB port - Maybe @eMadman can work something out for a replacement for you...Thanks for your help, sorry about the accident. 

I've tried uploading everything I could onto a test board and I can't seem to duplicate your locked html files.

**Reply by chrisbmcarthur** [reply]
LOL - I do appreciate the humour!
Using the URL on the smartspin2k SSID I was able to restore, program the home network SSID, and then get back in. The current "need" for a working USB port has been negated! The damage is purely my fault. The desire to use the newest code is due to my curious personality (others may use a different term or description).
I can send you a zip of the files I uploaded....but I also dropped in a firmware.bin file as well to do a downgrade. The downgrade worked, but did it use up additional file space? My BIG learning is that I needed to pull a fresh GIT clone and then push the files from the clone...NOT - from the website (was on a PC that did not have Git/VCode).
With the shifter reset clearing partitions and all things "customized" by the user...this should be much easier for others.
Also - the OTA page - when the firmware upgrade is done..it just sits there....and the user is not sure what is occurring. Either pop a message, or put some further details on the screen to guide the user?
Thanks again - over and out for tonight.....seriously.

**Comment by VUTSHYK** [comment]
Hello.
I have a cheap turbo trainer tt-01. And I want to upgrade it using ss2k. I made my own gearbox for this. And I had a question whether 2 full rotations of the stepping motor are enough for the normal operation of the device.
![Знімок екрана 2024-01-15 084124](https://github.com/doudar/SmartSpin2k/assets/53313562/63b160b5-74dd-420e-8244-8c7e5a614ccb)

**Reply by doudar** [reply]
Yes, as long as the torque is low enough, it's not an issue but the torque needs to be very low.

**Reply by VUTSHYK** [reply]
Low torque will work on the trainer. Will I need to change settings in settings.h?
I need much fewer steps than on a spin bike.

**Reply by doudar** [reply]
You could increase incline multiplier and shiftstep greatly, but if you're good with re-compiling and using your own custom firmware, I'd recommend increasing this: 
https://github.com/doudar/SmartSpin2k/blob/c75f2e3ad5319e29fbdd0946b2caa2757839ffaa/src/Main.cpp#L450

**Reply by VUTSHYK** [reply]
That is, adjust the driver for more microsteps.

**Reply by doudar** [reply]
Exactly. That way it should keep the background math more in line

**Reply by VUTSHYK** [reply]
Please let me know.  5 is 1/16, 6 is 1/32, 7 is 1/64?

**Reply by doudar** [reply]
It's directly the number. I believe the comment in the code is currently incorrect. See below from the tmc stepper library: 

![image](https://github.com/doudar/SmartSpin2k/assets/17362216/4c38dbdf-c620-4bd5-8554-04bd012d40b0)

**Reply by doudar** [reply]
So I'd try 8 or maybe even 16

**Comment by sergeleo** [comment]
Hello everyone! I recently got SS2K for my Bowflex C6 and overall I'm very happy with it. Makes my training more enjoyable, especially when I don't have to worry about adjusting the resistance. However, it does glitch sometimes, I was following WiKis guide for setting it up for C6, and noticed when my wife spins, it often turns the knob to 100. Very annoying. She's a small Asian girl, and obviously she rides in low watts threshold, like 60 to 120 Watts, with 50-70 rpms. Are there any tweaks that I should consider?

**Reply by doudar** [reply]
My guess is it might have something to do with the low cadence. Below 70rpm, the algorithm for ERG isn't getting much information and so if she's doing an ERG workout, it might calculate a target point incorrectly. It would be interesting to switch up the cadence requirements, recompile and see if it has a positive effect. Try decreasing the following to 70 and recompile.
https://github.com/doudar/SmartSpin2k/blob/c75f2e3ad5319e29fbdd0946b2caa2757839ffaa/include/settings.h#L273

**Reply by sergeleo** [reply]
Awesome! Thank you for the tip! I'll do that and will let you know.

**Reply by doudar** [reply]
Awesome! Hopefully that helps.
